# Filtering chromosomes to include only biallelic SNPS
___
## Purpose of Repo:
This repo was created to familiarize myself with and re-create commands that can filter gzipped vcf files,
and compare the results.
___

## Script Descriptions:

## `biallelic.py`
#### INPUT:
1) A gzipped VCF file that is to be filtered
2) A VCF filename that will be the destination for the filtered data
#### OUTPUT:
- A VCF file that contains only the biallelic snps from the input gzipped VCF file, preceded by the same header
#### FUNCTION:
- Filters a gzipped vcf file into a new gzipped vcf file that only contains biallelic SNPs
#### COMMAND EXAMPLE:
```python 
python biallelic.py vcf_input vcf_output
```
___
## `bcf_wrapper.sh`
This shell script contains commands for extracting biallelic sites from reference
chromosomes using three methods, 
recording and storing all indel sites that may have been left over in the filtered
chromosomes inside a text file,
and comparing text files generated by different methods to record discrepancies.

All commands are currently commented out but in the correct order.
Uncommenting each command and running the wrapper
(once the proper directory structure is created)
would replicate my work.


- **Command 1:** loads necessary modules with
`module load bcftools/1.13 gsl/2.5 gcc/8.3 perl/5.24.1`


- Note: Commands 2 and 3 use `bcftools` to filter the chromosome with different syntax. 
  In both commands, `-m2 -M2` indicates a minimum and maximum of two versions of the site. 
  The `-Oz` indicates a zipped output, while the `-o` precedes and indicates the output filepath.


- **Command 2:**
  ```bash
  bcftools view -m2 -M2 -i VT="SNP" -Oz -o iinfo_CHR.vcf.gz source_CHR.vcf.gz
  ```
  The `-i` calls to include sites for which the expression `VT="SNP"` is `True`.
  This is another way to avoid including indels, which would appear in the `INFO` column as `VT="INDEL"`.


- **Command 3:**
  ```bash
  bcftools view -m2 -M2 -v snps -Oz -o vsnp_CHR.vcf.gz source_CHR.vcf.gz
  ```
  The `-v` or `--variants only` makes sure only variant sites are output.
  Appending `-v snps` to `-m2 -M2` makes sure the output only displays biallelic snps.
  

- **Command 4:**
    ```bash
    python biallelic.py source_CHR.vcf.gz homebrew_CHR.vcf.gz
    ```
    Following the input format I wrote for `biallelic.py`, the first argument
    is the reference genome and the second argument is the filtered output filepath


- **Command 5:**
    ```bash
    python indel_pos.py vsnp_CHR.vcf.gz vsnps_CHR.txt
    ```
    Here, the command is simplified to not include long naming conventions and
    the specific filepath, but these can be found by looking inside `bcf_wrapper.sh`
    For organization's sake, the .txt file output of this and the following command
    (which are intended to be empty) are rerouted to their own directory, named
    `vsnp_indelpos` and `info_indelpos`, respectively.


- **Command 6:**
    ```bash
    python indel_pos.py info_CHR.vcf.gz info_CHR.txt
    ```
  

- **Command 7:**
    ```bash
    python indel_diff.py vsnps_CHR.txt info_CHR.txt indel_diff_CHR.txt
    ```
    This command is also simplified to avoid long filepaths. It calls `indel_diff.py`
    to compare the indel positions found in the vsnp and indel text output versions.
    The third argument is a filepath that includes the indel difference directory,
    which I called `diff_results`

___
## `compare_vcf.py`
#### INPUT:
1) Filepath to a gzipped VCF file that was the filtered result of my `biallelic.py` function
2) Filepath to a gzipped VCF file that was output by one of the `bcftools` biallelic filtering commands found in `biallelic.sh`
#### OUTPUT:
- Prints `identical`, a boolean which indicates whether the gzipped input vcf files are identical.
 Notably, these files are expected to differ in their headers, where the name of the function with which they were created is recorded.
 This function explicitly ignores that difference.
- If a discrepancy is found, the line at which they begin to differ is indicated with a print statement.
  The function ends after the first discrepancy is found.
#### PURPOSE:
Returns whether two vcf files (such as one produced by `biallelic.py` and one by `bcftools` filtering) are identical to each other
#### COMMAND EXAMPLE:
```python
python compare_vcf my_biallelic.vcf.gz bcftools_biallelic.vcf.gz
```
___
## `indel_pos.py`
#### INPUT:
- A gzipped vcf file representing a chromosome
#### OUTPUT:
- Filepath to a .txt output that contains the indel positions from the vcf file.
  The chromosome reference number (ie. 1, 2, 3, etc., not chr1, chr2, chr3, etc.) in column 1.
  Tab-delineated. The position of the single indel is in column 2, in the following tab-delineated column format: `CHROM   |   POS`
#### PURPOSE:
Takes in a gzipped vcf file (such as one produced by filtering out indels with `bcftools`)
and writes all indels present in the file to a new text file.
#### COMMAND EXAMPLE:
```python
python indel_pos.py filtered_CHR.vcf.gz out.txt
```
___
## `indel_diff.py`
#### INPUT:
1) Filepath to a .txt output generated by filtering with the `-v snps` `bcftools` command
2) Filepath to a .txt output generated by filtering with the `-i VT="SNP"` `bctools` command
#### OUTPUT:
- Filepath to a .txt output that contains all examples in either input that is not present in the other file, in the following format: 
`CHROM | VERSION (-v or -i) | POS`
#### PURPOSE:
Takes in two text files that resulted from `indel_pos.py` and compares them to look for discrepancies.
If there are any discrepancies in the text files, the text file output records them.
#### COMMAND EXAMPLE:
```python
python indel_diff.py vsnp_CHR.txt iinfo_CHR.txt indel_diff_CHR.txt
```
___
___
## Directory Description:
- `vsnp_indelpos` : a destination for the output of `indel_pos.py` performed on the gzipped vcf output given by the `-v snps` `bcftools` command.
- `info_indelpos` : a destination for the output of `indel_pos.py` performed on the gzipped vcf output given by the `-i VT="SNP"` `bcftools` command.
- `diff_results` : a destination for the output of `indel_diff.py` performed on .txt files found in `vsp_indelpos` and `info_indelpos` for each chromosome.
- `diff_tests` : a destination for test inputs and test cases for the function `indel_diff.py`
___
___
## Step-by-Step Analysis Walkthrough:

### Shell Scripts vs the Command Line:
The simplest way to execute all these steps would be to uncomment and run each command listed in 
`bcf_wrapper.sh`. However, in this walkthrough I will be demonstrating the process using bash command-line
job submissions, with full filepaths and names included.

### Bash Command Structure:

```bash
for CHR in {1..22};
do sbatch -J ${CHR}_tgp_filter -N 1 -n 4 -t 1:00:00 --mem=2G
-o logs/${CHR}_tgp_filter-%A.out -e logs/${CHR}_tgp_filter-%A.err
--mail-type=ALL --mail-user=brian_kirz@brown.edu
--wrap="module load bcftools/1.13 gsl/2.5 gcc/8.3 perl/5.24.1; bcftools view -m2 -M2 -v snps -Oz -o vsnp_biallelic_snps_chr${CHR}.vcf.gz /users/bkirz/data/data/modern_genomes/1000genomes/ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr${CHR}.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz"; done
```
I'll be using this format often in this guide to submit jobs from the command line.
Here are some things to recognize about the command structure:

  ```bash
  for CHR in {1..22}; do
  ```
- This part of the command indicates that the job should be run on each chromosme,
  using the CHR variable as a standin to indentify other files that should be treated.
```bash
sbatch -J ${CHR}_tgp_filter -N 1 -n 4 -t 1:00:00 --mem=2G
```
- This line indicates some constant parameters: the system's job name, the number of cores,
  ensuring that all cores are on one machine, runtime in minutes, and memory.
```bash
-o logs/${CHR}_tgp_filter-%A.out -e logs/${CHR}_tgp_filter-%A.err
```
- Indicates where the job and error files should go.
```bash
--mail-type=ALL --mail-user=brian_kirz@brown.edu
```
- Sets my email as a notification destination.
```bash
--wrap="module load bcftools/1.13 gsl/2.5 gcc/8.3 perl/5.24.1; bcftools view -m2 -M2 -v snps -Oz -o vsnp_biallelic_snps_chr${CHR}.vcf.gz /users/bkirz/data/data/modern_genomes/1000genomes/ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr${CHR}.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz"; done
```
- This line will be familiar: it's the commands in `bcf_wrapper.sh`, semicolon-delinated
  and "wrapped" with  `--wrap="<Command Here>"; done`


### I. Filtering Chromosomes
First, you'll want to extract biallelic sites only from all 22 reference chromosomes using
`bcftools`, a builtin command, or `biallelic.py`, which I programmed.
The resulting chromosome (which were output into my main folder) should only contain biallelic sites.

The filepath I used for this is:
`/users/bkirz/data/data/modern_genomes/1000genomes/ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/`

All chromosomes I used are found in this directory and have the naming convention:
`ALL.chr<chr#>.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz`


#### 1. Using `bcftools` with vsnp:
```bash
for CHR in {1..22}; do
sbatch -J ${CHR}_bcftools_vsnp -N 1 -n 4 -t 1:00:00 --mem=2G
-o logs/${CHR}_bcftools_vsnp-%A.out -e logs/${CHR}_bcftools_vsnp-%A.err
--mail-type=ALL --mail-user=brian_kirz@brown.edu
--wrap="module load bcftools/1.13 gsl/2.5 gcc/8.3 perl/5.24.1; bcftools view -m2 -M2 -v snps -Oz -o vsnp_biallelic_snps_chr${CHR}.vcf.gz /users/bkirz/data/data/modern_genomes/1000genomes/ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr${CHR}.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz"; done
```

#### 2. Using `bcftools` with info:
```bash
for CHR in {1..22}; do
sbatch -J ${CHR}_bcftools_info -N 1 -n 4 -t 1:00:00 --mem=2G
-o logs/${CHR}_bcftools_info-%A.out -e logs/${CHR}_bcftools_info-%A.err
--mail-type=ALL --mail-user=brian_kirz@brown.edu
--wrap="module load bcftools/1.13 gsl/2.5 gcc/8.3 perl/5.24.1; bcftools view -m2 -M2 -i 'INFO/VT="SNP"' -Oz -o info_tgp_biallelic_snps_chr${CHR}.vcf.gz /users/bkirz/data/data/modern_genomes/1000genomes/ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr${CHR}.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz"; done
```

#### 3. Using `bialellic.py`
```bash
for CHR in {1..22}; do
sbatch -J ${CHR}_my_biallelic -N 1 -n 4 -t 1:00:00 --mem=2G
-o logs/${CHR}_my_biallelic-%A.out -e logs/${CHR}_my_biallelic-%A.err
--mail-type=ALL --mail-user=brian_kirz@brown.edu
--wrap="python biallelic.py /users/bkirz/data/data/modern_genomes/1000genomes/ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr${CHR}.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz my_biallelic_snps_chr${CHR}.vcf.gz"; done
```

### II. Comparing Filtered Chromosomes

#### 4. Comparing the results of `biallelic.py` and `bcftools -v snps`
```bash
for CHR in {1..22}; do
sbatch -J ${CHR}_compare_my_vsnp -N 1 -n 4 -t 1:00:00 --mem=2G
-o logs/${CHR}_compare_my_vsnp-%A.out -e logs/${CHR}_compare_my_vsnp-%A.err
--mail-type=ALL --mail-user=brian_kirz@brown.edu
--wrap="python compare_vcf.py vsnp_biallelic_snps_chr${CHR}.vcf.gz my_biallelic_snps_chr${CHR}.vcf.gz"; done
```

#### 5. Comparing the results of `biallelic.py` and `bcftools -i VT="SNP"`
```bash
for CHR in {1..22}; do
sbatch -J ${CHR}_compare_my_info -N 1 -n 4 -t 1:00:00 --mem=2G
-o logs/${CHR}_compare_my_info-%A.out -e logs/${CHR}_compare_my_info-%A.err
--mail-type=ALL --mail-user=brian_kirz@brown.edu
--wrap="python compare_vcf.py info_biallelic_snps_chr${CHR}.vcf.gz my_biallelic_snps_chr${CHR}.vcf.gz"; done
```

### III. Extracting Indel Positions
Running `indel_pos.py` and putting the resulting .txt file of indel positions in the correct directory:

#### 6. Finding indel positions in the filtered output of `bcftools -v snps`
```bash
for CHR in {1..22}; do
sbatch -J ${CHR}_vsnp_indelpos_txt -N 1 -n 4 -t 1:00:00 --mem=2G 
-o logs/${CHR}_vsnp_indelpos_txt-%A.out -e logs/${CHR}_vsnp_indelpos_txt-%A.err
--mail-type=ALL --mail-user=brian_kirz@brown.edu
--wrap="python indel_pos.py vsnp_biallelic_snps_chr${CHR}.vcf.gz /users/bkirz/scratch/kirz_tutorials/vcf_processing/vsnp_indelpos/vsnps_chr${CHR}.txt"; done
```
#### 7. Finding indel positions in the filtered output of `bcftools -i VT="SNP"`
```bash
for CHR in {1..22}; do
sbatch -J ${CHR}_info_indelpos_txt -N 1 -n 4 -t 1:00:00 --mem=2G 
-o logs/${CHR}_info_indelpos_txt-%A.out -e logs/${CHR}_info_indelpos_txt-%A.err
--mail-type=ALL --mail-user=brian_kirz@brown.edu
--wrap="python indel_pos.py info_biallelic_snps_chr${CHR}.vcf.gz /users/bkirz/scratch/kirz_tutorials/vcf_processing/info_indelpos/info_chr${CHR}.txt"; done
```
#### 8. Finding indel positions in the filtered output of `biallelic.py`
```bash
for CHR in {1..22}; do
sbatch -J ${CHR}_my_indelpos_txt -N 1 -n 4 -t 1:00:00 --mem=2G 
-o logs/${CHR}_my_indelpos_txt-%A.out -e logs/${CHR}_my_indelpos_txt-%A.err
--mail-type=ALL --mail-user=brian_kirz@brown.edu
--wrap="python indel_pos.py my_biallelic_snps_chr${CHR}.vcf.gz /users/bkirz/scratch/kirz_tutorials/vcf_processing/my_indelpos/info_chr${CHR}.txt"; done
```

### IV. Comparing Indel Positions
#### 9. Running `indel_diff.py` on the `indel_pos.py` outputs of both `bcftools` methods
```bash
for CHR in {1..22}; do
sbatch -J ${CHR}_indel_diff -N 1 -n 4 -t 1:00:00 --mem=2G 
-o logs/${CHR}_indel_diff-%A.out -e logs/${CHR}_indel_diff-%A.err 
--mail-type=ALL --mail-user=brian_kirz@brown.edu 
--wrap="python indel_diff.py vsnp_indelpos/vsnps_chr${CHR}.txt info_indelpos/info_chr${CHR}.txt /users/bkirz/scratch/kirz_tutorials/vcf_processing/diff_results/indel_diff_chr${CHR}.txt"; done
```